Lib/test/test_capi/test_tuple.py

```diff
@@ -0,0 +1,206 @@
+
+        self.assertRaises(IndexError, getitem, tup, PY_SSIZE_T_MIN)
+        self.assertRaises(IndexError, getitem, tup, -1)
+        self.assertRaises(IndexError, getitem, tup, len(tup))
+        self.assertRaises(IndexError, getitem, tup, PY_SSIZE_T_MAX)
+        self.assertRaises(SystemError, getitem, [1, 2, 3], 1)
+        self.assertRaises(SystemError, getitem, 42, 1)
+
+        # CRASHES getitem(NULL, 1)
+
+    def test_tuple_get_item(self):
+        # Test PyTuple_GET_ITEM()
+        get_item = _testcapi.tuple_get_item
+
+        tup = TupleSubclass((1, 2, 3))
+        self.assertEqual(get_item(tup, 0), 1)
+        self.assertEqual(get_item(tup, 2), 3)
+
+        tup = (1, 2, 3)
+        self.assertEqual(get_item(tup, 0), 1)
+        self.assertEqual(get_item(tup, 2), 3)
+
+    def test_tuple_getslice(self):
+        # Test PyTuple_GetSlice()
+        getslice = _testlimitedcapi.tuple_getslice
+
+        # empty
+        tup = (1, 2, 3)
+        self.assertEqual(getslice(tup, PY_SSIZE_T_MIN, 0), ())
+        self.assertEqual(getslice(tup, -1, 0), ())
+        self.assertEqual(getslice(tup, 3, PY_SSIZE_T_MAX), ())
+        self.assertEqual(getslice(tup, 0, 0), ())
+        self.assertEqual(getslice(tup, 3, 0), ())
+        tup = TupleSubclass((1, 2, 3))
+        self.assertEqual(getslice(tup, PY_SSIZE_T_MIN, 0), ())
+        self.assertEqual(getslice(tup, -1, 0), ())
+        self.assertEqual(getslice(tup, 3, PY_SSIZE_T_MAX), ())
+        self.assertEqual(getslice(tup, 0, 0), ())
+        self.assertEqual(getslice(tup, 3, 0), ())
+
+        # slice
+        tup = (1, 2, 3)
+        self.assertEqual(getslice(tup, 1, 3), (2, 3))
+        tup = TupleSubclass((1, 2, 3))
+        self.assertEqual(getslice(tup, 1, 3), (2, 3))
+
+        # whole
+        tup = (1, 2, 3)
+        self.assertEqual(getslice(tup, 0, 3), tup)
+        self.assertEqual(getslice(tup, 0, 100), tup)
+        self.assertEqual(getslice(tup, -100, 100), tup)
+        tup = TupleSubclass((1, 2, 3))
+        self.assertEqual(getslice(tup, 0, 3), tup)
+        self.assertEqual(getslice(tup, 0, 100), tup)
+        self.assertEqual(getslice(tup, -100, 100), tup)
+
+        self.assertRaises(SystemError, getslice, [1, 2, 3], 0, 1)
+        self.assertRaises(SystemError, getslice, 42, 0, 1)
+        # CRASHES getslice(NULL, 0, 0)
+
+    def test_tuple_setitem(self):
+        # Test PyTuple_SetItem()
+        setitem = _testlimitedcapi.tuple_setitem
+
+        tup = (0, 0)
+        self.assertEqual(setitem(tup, 0, 1), (1, 0))
+        self.assertEqual(setitem(tup, 1, 1), (0, 1))
+
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MIN, 1)
+        self.assertRaises(IndexError, setitem, tup, -1, 1)
+        self.assertRaises(IndexError, setitem, tup, len(tup), 1)
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MAX, 1)
+        self.assertRaises(SystemError, setitem, [1], 0, 1)
+        self.assertRaises(SystemError, setitem, 42, 0, 1)
```

Comment:
These errors are raised not by the tested function, but by the improper use of the C API in the wrapper. You should only make a new tuple if the argument is a tuple.

Reply:
Yep, fixed (with a slight modification of your proposal).